solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+    {
+        // The expected structure of the preimage created during signing of a
+        // non-witness input:
+        // - transaction version (4 bytes)
+        // - number of inputs as written as varint (1 byte, 3 bytes, 5 bytes or
+        //   9 bytes)
+        // - for each input
+        //   - outpoint (hash and index) (36 bytes)
+        //   - unlocking script for the input being signed (variable length)
+        //     or `00` for all other inputs (1 byte)
+        //   - input sequence (4 bytes)
+        // - outputs (variable length)
+        // - transaction locktime (4 bytes)
+        // - sighash type (4 bytes)
+
+        // The input data begins at the constant offset of 4 (the first 4 bytes
+        // are for the transaction version).
+        (uint256 inputsCompactSizeUintLength, uint256 inputsCount) = preimage
+            .parseVarIntAt(4);
+
+        // To determine the first input starting index, we must jump 4 bytes
+        // over the transaction version length and the compactSize uint which
+        // prepends the input vector. One byte must be added because
+        // `BtcUtils.parseVarInt` does not include compactSize uint tag in the
+        // returned length.
+        //
+        // For >= 0 && <= 252, `BTCUtils.determineVarIntDataLengthAt`
+        // returns `0`, so we jump over one byte of compactSize uint.
+        //
+        // For >= 253 && <= 0xffff there is `0xfd` tag,
+        // `BTCUtils.determineVarIntDataLengthAt` returns `2` (no
+        // tag byte included) so we need to jump over 1+2 bytes of
+        // compactSize uint.
+        //
+        // Please refer `BTCUtils` library and compactSize uint
+        // docs in `BitcoinTx` library for more details.
+        uint256 inputStartingIndex = 4 + 1 + inputsCompactSizeUintLength;
+
+        for (uint256 i = 0; i < inputsCount; i++) {
+            uint256 inputLength = preimage.determineInputLengthAt(
+                inputStartingIndex
+            );
+
+            bytes memory input = preimage.slice(
+                inputStartingIndex,
+                inputLength
+            );
+
+            bytes32 previousOutpointTxIdLe = input.extractInputTxIdLE();
+            uint32 previousOutpointIndex = BTCUtils.reverseUint32(
+                uint32(input.extractTxIndexLE())
+            );
+
+            bytes memory scriptSig = input.extractScriptSig();
+
+            if (scriptSig.length > 1) {
+                // The input this preimage was generated for was found.
+                // All the other inputs in the preimage are marked with a null
+                // scriptSig ("00") which has length of 1.
+                utxoKey = uint256(
+                    keccak256(
+                        abi.encodePacked(
+                            previousOutpointTxIdLe,
+                            previousOutpointIndex
+                        )
+                    )
+                );
+
+                break;
+            }
```

Comment:
I think we can do it more effectively:
```
            uint256 inputLength = preimage.determineInputLengthAt(
                inputStartingIndex
            );

            (, uint256 scriptSigLength) = preimage.extractScriptSigLenAt(inputStartingIndex);

            if (scriptSigLength > 1) {
                // The input this preimage was generated for was found.
                // All the other inputs in the preimage are marked with a null
                // scriptSig ("00") which has length of 1.
                bytes32 outpointTxHash = preimage.extractInputTxIdLeAt(inputStartingIndex);
                uint32 outpointIndex = BTCUtils.reverseUint32(
                    uint32(preimage.extractTxIndexLeAt(inputStartingIndex))
                );

                utxoKey = uint256(
                    keccak256(
                        abi.encodePacked(
                            outpointTxHash, 
                            outpointIndex
                        )
                    )
                );

                break;
            }
```
Would it work?

Reply:
Done in e7f7d6fb8100a057dbee2b281b4de467f0feb938.